Source/Core/Core/IOS/USB/Emulated/Skylanders/SkylanderFigure.cpp

```diff
@@ -0,0 +1,445 @@
+
+  memcpy(hashIn.data(), dest, 0x56);
+}
+void SkylanderFigure::Encrypt(u8 input[FIGURE_SIZE])
+{
+  std::array<u8, 0x56> hashIn = {};
+
+  GenerateIncompleteHashIn(hashIn.data());
+
+  u8 encrypted[FIGURE_SIZE] = {};
+
+  u8 currentBlock[0x10] = {};
+
+  // Run for every block
+  for (u8 i = 0; i < 64; i++)
+  {
+    memcpy(currentBlock, input + (i * 0x10), 0x10);
+
+    // Skip sector trailer and the first 8 blocks
+    if (((i + 1) % 4 == 0) || i < 8)
+    {
+      memcpy(encrypted + (i * 0x10), currentBlock, 0x10);
+      continue;
+    }
+
+    // Block index
+    hashIn[0x20] = i;
+
+    std::array<u8, 0x10> hashOut = {};
+
+    mbedtls_md5_ret(hashIn.data(), 0x56, hashOut.data());
+
+    mbedtls_aes_context aes_context = {};
+
+    mbedtls_aes_setkey_enc(&aes_context, hashOut.data(), 128);
+
+    mbedtls_aes_crypt_ecb(&aes_context, MBEDTLS_AES_ENCRYPT, currentBlock, encrypted + (i * 0x10));
+  }
+
+  memcpy(data.data(), encrypted, FIGURE_SIZE);
+
+  NOTICE_LOG_FMT(IOS_USB, "Encrypted skylander data: \n{}", HexDump(encrypted, FIGURE_SIZE));
+}
+SkylanderFigure::SkylanderFigure(File::IOFile file)
+{
+  sky_file = std::move(file);
+  sky_file.Seek(0, File::SeekOrigin::Begin);
+  sky_file.ReadBytes(data.data(), data.size());
+}
+bool SkylanderFigure::Create(u16 m_sky_id, u16 m_sky_var)
+{
+  if (!sky_file)
+  {
+    return false;
+  }
+
+  const auto file_data = data.data();
+
+  memset(data.data(), 0, data.size());
+
+  PopulateSectorTrailers();
+
+  // Set the NUID of the figure
+  Common::Random::Generate(&file_data[0], 4);
+
+  // The BCC (Block Check Character)
+  file_data[4] = file_data[0] ^ file_data[1] ^ file_data[2] ^ file_data[3];
+
+  // ATQA
+  file_data[5] = 0x81;
+  file_data[6] = 0x01;
+
+  // SAK
+  file_data[7] = 0x0F;
+
+  // Set the skylander info
+  memcpy(&file_data[0x10], &m_sky_id, sizeof(m_sky_id));
+  memcpy(&file_data[0x1C], &m_sky_var, sizeof(m_sky_var));
+
+  // Set checksum
+  u16 checksum = SkylanderCrypto::ComputeCRC16(0xFFFF, file_data, 0x1E);
+  memcpy(&file_data[0x1E], &checksum, sizeof(checksum));
+
+  PopulateKeys();
+
+  Save();
+  return true;
+}
+void SkylanderFigure::Save()
+{
+  sky_file.Seek(0, File::SeekOrigin::Begin);
+  sky_file.WriteBytes(data.data(), 0x40 * 0x10);
+}
+
+void SkylanderFigure::GetBlock(u8 index, u8* dest) const
+{
+  memcpy(dest, data.data() + (index * 0x10), 0x10);
+}
+
+FigureData SkylanderFigure::GetData() const
+{
+  FigureData figureData = {.figureID = PointerToU16(data.data() + 0x10),
+                           .variantID = PointerToU16(data.data() + 0x1C)};
+
+  auto filter = std::make_pair(figureData.figureID, figureData.variantID);
+  Type type = Type::Item;
+  if (IOS::HLE::USB::list_skylanders.count(filter) != 0)
+  {
+    auto found = IOS::HLE::USB::list_skylanders.at(filter);
+    type = found.type;
+  }
+
+  figureData.normalizedType = (u8)NormalizeSkylanderType(type);
+  if ((Type)figureData.normalizedType == Type::Skylander)
```

Comment:
`static_cast` rather than C style casts.

```cpp
static_cast<u8>(NormalizeSkylanderType(type));
static_cast<Type>(figureData.normalizedType);
```